### PR TITLE
migrate PO00{1,2,3,4,5} to ST00{3,4,5,6,7}

### PR DESCRIPTION
**Pull request by DinoChiesa**
This fixes issue #339 

- To consider Condition placements, and to report line numbers on the endpoint file, these checks should be on the Step, not on the Policy, and should use onProxyEndpoint and onTargetEndpoint.  Therefore move PO00{1,2,3,4,5} to ST00{3,4,5,6,7}, and re-factor those plugins so they report on the Endpoints.

- Also introduce new tests for these cases, for Sharedflows.

- And modify the Endpoint module so that it correctly reports errors in Sharedflows. (This never worked!)  And introduce new tests to verify that this is working correctly. With this change, I see that EP002 and PD005 are incorrectly reporting errors on Step elements within SharedFlows.  I will fix that in a subsequent PR.

### CODE REVIEW

**Comment by DinoChiesa** [comment]
Also see the PR #343

File: test/fixtures/resources/ThreatProtection-Attachment/apiproxy/proxies/endpoint1.xml [hunk 1]
```diff
@@ -0,0 +1,116 @@
+<ProxyEndpoint name="endpoint1">
+  <Description>Proxy Endpoint 1</Description>
+  <HTTPProxyConnection>
+    <BasePath>/tp-attachment</BasePath>
+    <Properties/>
+    <VirtualHost>secure</VirtualHost>
+  </HTTPProxyConnection>
+
+  <FaultRules/>
```
Comment (ssvaidyanathan) [inline test/fixtures/resources/ThreatProtection-Attachment/apiproxy/proxies/endpoint1.xml:9]: We should include a test case with EV policy in the FaultRule section

File: lib/package/plugins/_extractVariablesPayloadCheck.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+  check(endpoint) {
+    let debug = this.debug;
+    this.debug(`check (${endpoint.getType()})`);
+    let reKnownMessages = "^(request|response|message)$";
+    let flagged = false;
+    let bundlePolicies = endpoint.parent.getPolicies();
+    let checker = this;
+
+    endpoint.getSteps().forEach( step => {
```
[the remark this answers is not in the document]
Comment (kurtkanaskie) [reply lib/package/plugins/_extractVariablesPayloadCheck.js:42]: It does catch those in FaultRules and DefaultFaultRule.

[the remark this answers is not in the document]
Comment (ssvaidyanathan) [reply lib/package/plugins/_extractVariablesPayloadCheck.js:42]: Thanks for confirming. Will mark this as resolved

File: lib/package/plugins/_policyConditionCheck.js [hunk 3]
```diff
@@ -14,52 +14,89 @@
...
+const stringNonNullCheckRegexp = (source) => `\\(? *\\b${source}\\b *(!=|NotEquals|IsNot) *(\"\"|null) *\\)?`;
+
+const messageCheckRegexp = (source) => stringNonNullCheckRegexp(`${source}.content`);
+
+const checkFactory = (plugin, policyType, debug) =>
+  function(endpoint) {
+    let flagged = false;
+    let bundlePolicies = endpoint.parent.getPolicies();
+
+    endpoint.getSteps().forEach( step => {
```
[the remark this answers is not in the document]
Comment (kurtkanaskie) [reply lib/package/plugins/_policyConditionCheck.js:28]: It does catch those in FaultRules and DefaultFaultRule.

**Comment by kurtkanaskie** [comment]
Given a policy with:
```
<ExtractVariables continueOnError="false" enabled="true" name="Extract-Variables-FormParam">
  <DisplayName>Extract Variables-FormParam</DisplayName>
  <FormParam name="greeting">
    <Pattern>hello {user}</Pattern>
  </FormParam>
  <Variable name="name"/>
  <IgnoreUnresolvedVariables>true</IgnoreUnresolvedVariables>
</ExtractVariables>
```
I get 3 errors - ST005 - for the steps using Extract-Variables-FormParam policy.
I don't get those errors in the version 2.30.
```
  <PreFlow name="PreFlow">
    <Request>
      <Step>
        <Condition>(request.content != null)</Condition>
        <Name>Extract-Variables-1</Name>
      </Step>
      <Step>
        <Condition>(request.formparams.count != 0)</Condition>
        <Name>Extract-Variables-FormParam</Name>
      </Step>
      <Step>
        <Condition>(request.formparams.greeting)</Condition>
        <Name>Extract-Variables-FormParam</Name>
      </Step>
      <Step>
        <Condition>(request.content != null)</Condition>
        <Name>Extract-Variables-FormParam</Name>
      </Step>
      <Step>
        <Condition>(request.content != null)</Condition>
        <Name>Extract-Variables-1</Name>
      </Step>
    </Request>
    <Response/>
  </PreFlow>
  ```

File: lib/package/plugins/_policyConditionCheck.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (DinoChiesa) [reply lib/package/plugins/_policyConditionCheck.js:28]: Yes. I have updated the tests to include tests for Steps that appear within FaultRules.

File: test/fixtures/resources/ThreatProtection-Attachment/apiproxy/proxies/endpoint1.xml (hunk 1 above)
Comment (DinoChiesa) [reply test/fixtures/resources/ThreatProtection-Attachment/apiproxy/proxies/endpoint1.xml:9]: Done.

**Comment by DinoChiesa** [comment]
I have modified ST005 , so that the condition must be one of these:

1. `request.formstring != null`
2. `request.formparam.PARAMNAME != null`, where `PARAMNAME` is the name cited in the policy in the `FormParam` element.

If there are multiple `FormParam` elements, then the plugin checks for a condition that checks at least one of them.

**Comment by DinoChiesa** [comment]
More changes coming. . . . .

**Comment by kurtkanaskie** [comment]
This took me awhile to find, given a policy without a Source element (optional - default uses message) or with a Source element that is anything but `<Source>request</Source>` I get insufficient Condition.
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<ExtractVariables continueOnError="false" enabled="true" name="Extract-Variables-FormParam">
  <DisplayName>Extract Variables-FormParam</DisplayName>
  <FormParam name="greeting">
    <Pattern>hello {user}</Pattern>
  </FormParam>
  <IgnoreUnresolvedVariables>true</IgnoreUnresolvedVariables>
</ExtractVariables>
```
These Conditions are insufficient without a Source element in the policy.
```|  <PreFlow name="PreFlow">
    <Request>
      <Step>
        <Condition>request.formstring != null</Condition>
        <Name>Extract-Variables-FormParam</Name>
      </Step>
      <Step>
        <Condition>request.formparam.greeting != null</Condition>
        <Name>Extract-Variables-FormParam</Name>
      </Step>
```

Easy to reproduce, just remove Source element from your test policy in `EV-Formparam-1.xml`

Also, I think we should add the test conditions to the warning message, it took me a while to figure out the right Condition.

Perhaps:
Use <Condition>request.formstring != null</Condition> or <Condition>request.formparam.PARAMNAME_FROM_POLICY != null</Condition>

**Comment by ssvaidyanathan** [comment]
@DinoChiesa - Should we add a note in the README about the plugin rule change from PO to ST?

**Pull request merged**